functionary/train/train.py

```diff
@@ -248,6 +265,55 @@ def train():
         print_rank0("***** HERE ARE SOME EXAMPLES FROM EVALUATION ***")
         training_utils.print_some_examples(eval_dataset, tokenizer)
 
+    # Dynamic batch size based on max tokens per batch
```

Comment:
While dynamic batch size might seem a good idea for dealing with memory issues, it would cause instabilities in training.  Due to the difference in gradient updates per conversation. I would rather prefer stability over memory efficiency. Stable updates with higher cost of GPUs should be preferred over a cheaper&faster training